system/jlib/jevent.cpp

```diff
@@ -930,75 +930,621 @@ bool readEvents(const char* filename, IEventVisitor& visitor)
+{
+    return new CPTreeEventVisitor();
+}
+
+// Concrete extension of TDumpEventVisitor<> that writes visited data to an output stream. One line
+// of text is produced for each event and attribute visited, plus additional lines for the file
+// name, version, and number of bytes read from the file.
+class CDumpTextEventVisitor : public TDumpEventVisitor<>
+{
+public:
+    virtual bool visitFile(const char* filename, uint32_t version) override
+    {
+        state = State::Header;
+        doVisitHeader(filename, version);
+        return true;
+    }
+
+    virtual Continuation visitEvent(EventType id) override
+    {
+        if (inHeader())
+        {
+            closeElement();
+            state = State::Events;
+        }
+        openElement(queryEventName(id));
+        doVisitEvent(id);
         return visitContinue;
     }
-    virtual Continuation visitAttribute(EventAttr id, uint32_t value) override
+
+    virtual bool departEvent() override
+    {
+        closeElement();
+        return true;
+    }
+
+    virtual void departFile(uint32_t bytesRead) override
     {
-        *out << "attribute: " << queryEventAttributeName(id) << " = " << value << std::endl;
+        if (inHeader())
+            closeElement();
+        state = State::Footer;
+        doVisitFooter(bytesRead);
+        closeElement();
+    }
+
+protected:
+    virtual void recordAttribute(const char* name, const char* value, bool quoted) override
+    {
+        markup.append("attribute: ").append(name).append(" = ");
+        if (quoted)
+            markup.append('\'');
+        markup.append(value);
+        if (quoted)
+            markup.append('\'');
+        markup.append('\n');
+    }
+
+    void dump(bool eoln)
+    {
+        out << markup.str();
+        markup.clear();
+        if (eoln)
+            out << std::endl;
+    }
+
+    bool inHeader() const { return State::Header == state; }
+    bool inEvents() const { return State::Events == state; }
+
+    void openElement(const char* name)
+    {
+        markup.append("event: ").append(name).append('\n');
+    }
+
+    void closeElement()
+    {
+        dump(false);
+    }
+
+public:
+    CDumpTextEventVisitor(std::ostream& _out)
+        : out(_out)
+    {
+    }
+
+protected:
+    enum class State : byte
+    {
+        Idle,
+        Header,
+        Events,
+        Footer,
+    };
+    State state{State::Idle};
+    std::ostream& out;
+    StringBuffer markup;
+};
+
+IEventVisitor* createDumpTextEventVisitor(std::ostream& out)
+{
+    return new CDumpTextEventVisitor(out);
+}
+
+// Extension of CDumpTextEventVisitor that outputs XML-formatted text. Header and Footer elements
+// synthesized to hold the file name, version, and number of bytes read from the file.
+class CDumpXMLEventVisitor : public CDumpTextEventVisitor
+{
+public:
+    virtual bool visitFile(const char* filename, uint32_t version) override
+    {
+        state = State::Header;
+        openElement(DUMP_STRUCTURE_ROOT, true);
```

Comment:
This sequence looks very similar in the json/xml/yaml implementaions.  Can it be in the base class?

Reply:
Possibly, but I'm not sure. I didn't want to get hung up on trying to unify these methods when it wasn't immediately apparent to me, unlike with visitAttribute.

Reply:
can the object/element terminology be commoned up, or is it not worth it?